job-server/test/spark.jobserver/LocalContextSupervisorSpec.scala

```diff
@@ -80,8 +82,8 @@ class LocalContextSupervisorSpec extends TestKit(LocalContextSupervisorSpec.syst
 
     it("should be able to add multiple new contexts") {
       supervisor ! AddContext("c1", contextConfig)
-      supervisor ! AddContext("c2", contextConfig)
       expectMsg(ContextInitialized)
+      supervisor ! AddContext("c2", contextConfig)
```

Comment:
Anyways, I would suggest, to unblock the 1.2 support, why don't we add a
comment around the switching of lines in the test, with a reference to the
SPARK-2243 and documentation.   The right way to address this is in Spark
itself; in the meantime let's merge this in to unlock others.  :)

Many thanks,

On Sun, Jan 18, 2015 at 12:26 AM, Evan Chan <velvia@gmail.com> wrote:

> But that's obviously not true. When you switched the lines in the test,
> you are not stopping one context before starting the other one right?   So
> that there are multiple contexts running at once.
>
>
>
> On Fri, Jan 16, 2015 at 11:46 AM, Fernando Otero <notifications@github.com
> > wrote:
>
>> In job-server/test/spark.jobserver/LocalContextSupervisorSpec.scala
>> <https://github.com/spark-jobserver/spark-jobserver/pull/39#discussion-diff-23104353>
>> :
>>
>> >        expectMsg(ContextInitialized)
>> > +      supervisor ! AddContext("c2", contextConfig)
>>
>> And I guess that I should have read the SparkContext documentation where
>> it says: * Only one SparkContext may be active per JVM. You must `stop()`
>> the active SparkContext before * creating a new one. This limitation may
>> eventually be removed; see SPARK-2243 for more details. so Obviously this
>> should fail.
>> … <#14afc263bf6d6469_14af4484df6e7bb5_>
>> On Fri, Jan 16, 2015 at 4:42 PM, Fernando O. <fotero@gmail.com> wrote: >
>> So I just wrote a simple test for spark and it seems like creating 2 >
>> contexts in parallel produces a failure: > > > test("Create 2 spark
>> contexts concurrently") { >
>> withSystemProperty("spark.driver.allowMultipleContexts", "true") { > var
>> secondSparkContext: SparkContext = null > var firstSparkContext:
>> SparkContext = null > try { > val pool = Executors.newFixedThreadPool(2) >
>> val future = new FutureTask[SparkContext](new Callable[SparkContext]() { >
>> def call(): SparkContext = { > val conf = new SparkConf ().setAppName
>> ("test").setMaster ("local") > sc = new SparkContext (conf) > new
>> SparkContext (conf) > }}) > > > val future2 = new
>> FutureTask[SparkContext](new Callable[SparkContext]() { > def call():
>> SparkContext = { > val conf = new SparkConf ().setAppName
>> ("test2").setMaster ("local") > sc = new SparkContext (conf) > new
>> SparkContext (conf) > > }}) > > pool.execute(future) >
>> pool.execute(future2) > > firstSparkContext = future.get() >
>> secondSparkContext = future2.get() > } finally { > Option
>> (secondSparkContext).foreach (_.stop () ) > Option
>> (firstSparkContext).foreach (_.stop () ) > } > } > } > > > fails with > >
>> akka.actor.InvalidActorNameException: actor name [LocalBackendActor] is >
>> not unique! > java.util.concurrent.ExecutionException: >
>> akka.actor.InvalidActorNameException: actor name [LocalBackendActor] is not
>> > unique! > at java.util.concurrent.FutureTask.report(FutureTask.java:122)
>> > at java.util.concurrent.FutureTask.get(FutureTask.java:188) > at >
>> org.apache.spark.SparkContextSuite$$anonfun$4$$anonfun$apply$mcV$sp$4.apply$mcV$sp(SparkContextSuite.scala:105)
>> > at >
>> org.apache.spark.SparkContextSuite$$anonfun$4$$anonfun$apply$mcV$sp$4.apply(SparkContextSuite.scala:80)
>> > > > > I'm sending an email to spark-users > > > > On Fri, Jan 16, 2015 at
>> 1:47 PM, Evan Chan <notifications@github.com> > wrote: > In
>> job-server/test/spark.jobserver/LocalContextSupervisorSpec.scala > <
>> https://github.com/spark-jobserver/spark-jobserver/pull/39#discussion-diff-23092200>
>> > : > > > expectMsg(ContextInitialized) > > + supervisor ! AddContext("c2",
>> contextConfig) > > I suspect this will take a while to investigate. I'd
>> like to understand > further why it doesn't work...... as a first pass, how
>> about posting the > logs from the test log when you don't switch the order
>> of the > initialization? Serializing is not hard, we set up a queue and
>> make it > event-based; ie when one gets an event from the context that the
>> > initialization is done, then we can create the second context. Let's
>> figure > out though why first. One initial problem was that creating
>> multiple > contexts at once was not possible due to global environment
>> vars, but that > problem has been solved already. > …
>> <#14af44476125e063_14af3a4d7cff36d0_> > On Fri, Jan 16, 2015 at 7:23 AM,
>> Fernando Otero <notifications@github.com> > wrote: In
>> job-server/test/spark.jobserver/LocalContextSupervisorSpec.scala < >
>> https://github.com/spark-jobserver/spark-jobserver/pull/39#discussion-diff-23085865>
>> > : > expectMsg(ContextInitialized) > + supervisor ! AddContext("c2", >
>> contextConfig) I included this change just to talk about it, since it >
>> actually seems like there's a race condition, I had no intentio to merge >
>> this change. — Reply to this email directly or view it on GitHub < >
>> https://github.com/spark-jobserver/spark-jobserver/pull/39/files#r23085865>
>> > . > -- The fruit of silence is prayer; the fruit of prayer is faith; the
>> > fruit of faith is love; the fruit of love is service; the fruit of
>> service > is peace. -- Mother Teresa > > — > Reply to this email directly
>> or view it on GitHub > <
>> https://github.com/spark-jobserver/spark-jobserver/pull/39/files#r23092200>
>> > . >
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/spark-jobserver/spark-jobserver/pull/39/files#r23104353>
>> .
>>
>
>
>
> --
> The fruit of silence is prayer;
> the fruit of prayer is faith;
> the fruit of faith is love;
> the fruit of love is service;
> the fruit of service is peace.  -- Mother Teresa
>



-- 
The fruit of silence is prayer;
the fruit of prayer is faith;
the fruit of faith is love;
the fruit of love is service;
the fruit of service is peace.  -- Mother Teresa